### ISSUE TITLE
unified image loading

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes-sigs/kind#2038]
As discussed when podman has come up previously I don't think we should continue making unique image loading commands, instead we should:

- support podman + docker + tarball in a single `kind load image`
- for each image argument, we should:
   - check for a file first if it looks like a path / check if the file exists
   - check docker
   - check podman
 
 For checking docker vs podman, we can potentially enhance it to cache if docker or podman are even in PATH and skip the ones that aren't installed.

We could also perhaps match the node provider in preferring these, but a couple things to consider:
- what about once we support another provider that may not have the ability to export images?
- what about predictable consistency across hosts?

I think it may be worth sticking to the simplest: a fixed order of these, and ensuring this is fast.
If a script/user _really_ want to load from a particular source, they can always do like `podman save hello-world:2 | kind load image -`

cc @aojea @amwat @bostrt

**Comment by aojea** [comment]
> For checking docker vs podman, we can potentially enhance it to cache if docker or podman are even in PATH and skip the ones that aren't installed.
> 

> We could also perhaps match the node provider in preferring these, but a couple things to consider:
> 
> * what about once we support another provider that may not have the ability to export images?
> * what about predictable consistency across hosts?

and adding the methods to the provider interface?
If a new provider doesn´t support it, it just throw an error

```go
type Image interface {
   Save(image, dest string) error
   ImageID(image string) (string,error)
```

**Comment by BenTheElder** [comment]
> and adding the methods to the provider interface?

No, because we shouldn't need a provider instance. Even if I use docker to build some image (maybe I'm using a project that uses buildkit / buildx) I should be able to load it into my podman cluster smoothly, and vice versa.

**Comment by aojea** [comment]
> If a script/user really want to load from a particular source, they can always do like podman save hello-world:2 | kind load image -

That is great, and is easy to implement https://github.com/kubernetes-sigs/kind/pull/2041

I like the piping approach and leverage other tools like  https://github.com/containers/skopeo to deal with the whole image format thing

**Comment by BenTheElder** [comment]
> We could also perhaps match the node provider in preferring these, but a couple things to consider:

Thinking about this some more, I think it would be reasonable to bump the current node provider to the 2nd preference (after checking for a file matching the argument) for each provider, just moving them and keeping the rest of the list in a fixed order. 

Then the results are predictable with future providers, but with each provider we know to have images, we can give preference for the matching image source.

I feel like as a user I would expect podman to have priority when using podman, and docker when using docker.
The order is going to be fairly arbitrary for ignite or any other future provider without this, but should be fixed and predictable / known.

**Comment by BenTheElder** [comment]
probably also consider OCI images from disk https://github.com/kubernetes-sigs/kind/issues/2626

**Comment by corinz** [comment]
+1 on this

**Comment by BenTheElder** [comment]
Note: `kind build node-image` now supports multiple input sources and seems to be working really well, I think we should take our experience there and develop `kind load image` with some automatic detection / defaulting of image source and some way for users to override the source.

It can work with (local) tarball files, docker/podman/nerdctl, defaulting to match the provider of the cluster being used, with some option to override the source detection.

I don't think we should support remote tarballs, at least for now, it increases the surface area for little gain, if you're using the network to import an image you should just pull it.

But supporting both tarballs and container images in one command that intelligently selects the source, with fallbacks (users can have multiple runtimes in use ...) and automatically matching the runtime being selected for the nodes makes sense.

I don't think this is terribly complicated, but it needs a little bit of spelling out the design, then implementing.

**Comment by BenTheElder** [comment]
Also, the reason to sort this out up front is because we're going to run out of good names for these commands, and we get one shot at introducing a new CLI interface before we either burn that name or have to break users (and we strongly prefer to not break users)

**Comment by BenTheElder** [comment]
cc @aojea @stmcginnis this is a recurring pain point, thoughts?

**Comment by stmcginnis** [comment]
I really like the idea of having a single `kind load image` command, regardless of the specific provider.

> It can work with (local) tarball files, docker/podman/nerdctl, defaulting to match the provider of the cluster being used, with some option to override the source detection.

+1!

**Comment by jberkus** [comment]
+1 on single `kind load image`  if it's reasonably possible.

**Comment by BenTheElder** [comment]
It should be, but we need someone to work on the design and implementation. Basic bugfix upkeep / issue triage and testing kubernetes take priority over podman local dev: https://kind.sigs.k8s.io/docs/contributing/project-scope/, and we've been staying pretty busy with those.

**Comment by Blaimi** [comment]
I use something like this:

```shell
export TAG="registry.gitlab.com/biletado/backend/assets/go:local:dev"
buildah bud --layers -t "$TAG" . \
&& kind load image-archive -n kind-cluster <(podman save "$TAG" --format oci-archive) \
&& kubectl rollout restart deployment assets
```

Simplification of this is IMHO not really useful, since you can't omit the `rollout restart` and to configure your Pod to have `imagePullPolicy: Never` which you don't want in production. If you use a local registry like proposed in #1213 and [here](https://kind.sigs.k8s.io/docs/user/local-registry/), you still need to copy the image. Even if the cluster can access the same images which are available on the host, provided by Podman or Docker and with `image PullPolicy: IfNotPresent` (which should be used in production only if you refer to an image hash which will not work in a dev environment like this), you have to configure your Pod in the cluster differently but would at least make the additional copy unnecessary.

What you want here is syntactic sugar for `kind load image-archive -n kind-cluster <(podman save "$TAG" --format oci-archive` with the problem that `kind` needs to understand different local implementations and configurations for local registries and their configurations. I don't think this feature is economical. An entry in the docs for the oneliner from above would be much more useful because it can be adapted for different machine configurations and costs no developer resources.

**Comment by BenTheElder** [comment]
> Simplification of this is IMHO not really useful, since you can't omit the rollout restart and to configure your Pod to have imagePullPolicy: Never which you don't want in production.

Eliminating that isn't the goal. The `kind load docker-image` is able to be a bit smarter than piping from `podman save` because we can at least optimize out two cases:
1. the image is loaded already with the correct tag
2. the image is loaded already but the tag changed

Versus
3. the image is not loaded  and we actually need to save and import it

For the first two cases, we can inspect image metadata more cheaply than actually saving it, which matters for large images.

Also, a lot of people find `kind load` convenient, so making that work in a more portable way is useful.

> If you use a local registry like proposed in https://github.com/kubernetes-sigs/kind/issues/1213 and [here](https://kind.sigs.k8s.io/docs/user/local-registry/), you still need to copy the image.

If you use a local registry it only needs to copy the changed layers.

> What you want here is syntactic sugar for kind load image-archive -n kind-cluster <(podman save "$TAG" --format oci-archive with the problem that kind needs to understand different local implementations and configurations for local registries and their configurations. 

This issue isn't related to registries though, it's related to importing images from files/tarballs.
For that, there are different performance concerns.

I don't understand this part of the comment, sugar over `kind load image-archive` to support different runtimes *is* the purpose of this issue. 

For local registries, we already have cross-runtime docs:
https://kind.sigs.k8s.io/docs/user/local-registry/

I would recommend using the local registry when possible. It has better performance characteristics for almost anything other than infrequently loading to a single node and is more portable to different tooling.

**Comment by Blaimi** [comment]
> > If you use a local registry like proposed in [#1213](https://github.com/kubernetes-sigs/kind/issues/1213) and [here](https://kind.sigs.k8s.io/docs/user/local-registry/), you still need to copy the image.
> 
> If you use a local registry it only needs to copy the changed layers.

Nope. In the given example you have actually another copy of the image. One on the host, one in the registry (the additional one) and one on every kind node.

There are two use cases which want or don't want to be addressed with this issue:
* commonly used development environment and other environments with access to some maintained registry like productive environments
* local (“local” as in “this workstation” or “my machine”!) development environments (you have a CI/CD-Pipeline? -> You have a registry!)

for the first case—having a registry—, hopefully your environment/company has some documentation and firewall rules and whatever to have, that this issue is irrelevant talking about.

for the second case, and that's what `kind` is made for: There are two roles on this topic:
* I, as a CI/CD expert want to have syntactic sugar on some oneliner, because I think it is nice to have everything working on my machine, but if it doesn't, I will file a bugreport for my very unique configuration, e.g. I want to have `kind load` to respect `CONTAINERS_MACHINE_PROVIDER` on my Linux from scratch host.
* I, as an application developer, want to ask my CI/CD expert to provide me a predefined development environment, where I can have an one-liner to rebuild and immediately test the software changes I didn't even commit.

I (really), as a CI/CD expert, product owner, and university lecturer, can only understand the requests of the second role, the app-developer role!

* As CI/CD expert: You, as a CI/CD expert, could find this Issue. You can provide the one-liner to the people requesting it. Providing this and different other commands as examples in the documentation could make it easier to find it for future CI/CD experts (commonly knows as AI).
* As product owner: This feature will cause way more issues than it will solve, especially if there is a so incredible simple workaround -> will never happen, uneconomical.
* As a lecturer: They don't fu*king care. They copy&paste everything you provide them. They would even install viruses if the command would be `sudo install dont_install_this_virus` as long as it is provided in a powerpoint.
* As a CI/CD expert in the role of an application developer: I don't care if the command my CI/CD expert gave me has 60 or 90 characters, at least I can document it in the DEVELOPERS_README.md. I started this readme six years ago when we swithed from Win8.1 to Win10. I'm pretty sure that I will refresh it, once my company starts switching to Win11 in November this year [sic!].

**Comment by BenTheElder** [comment]
> Nope. In the given example you have actually another copy of the image. One on the host, one in the registry (the additional one) and one on every kind node.

No, that's not correct ... the layers will be pushed to the registry on demand *and pulled on demand*, versus loading them into every node every time.

If you use `kind load image-archive` you are pushing every layer to all nodes every time, and you are duplicating those layers to all nodes. Workloads may not run on all nodes, and when they do, they typically re-use layers in iterative development.

> There are two use cases which want or don't want to be addressed with this issue:
> commonly used development environment and other environments with access to some maintained registry like productive environments

Usage in production is completely out of scope.

> local (“local” as in “this workstation” or “my machine”!) development environments (you have a CI/CD-Pipeline? -> You have a registry!)

**CI pipelines do not necessarily have a registry.** 
When we test Kubernetes itself, we avoid giving registry push access to PR testing ...

> for the second case, and that's what kind is made for: There are two roles on this topic:

... I might have an idea or two about what kind is made for. https://github.com/kubernetes-sigs/kind/pull/1

> I, as a CI/CD expert want to have syntactic sugar on some oneliner, because I think it is nice to have everything working on my machine, but if it doesn't, I will file a bugreport for my very unique configuration, e.g. I want to have kind load to respect CONTAINERS_MACHINE_PROVIDER on my Linux from scratch host.

Remember that podman support is experimental, hence the warning message on every log. Podman has been a moving target with breaking changes. 

> I, as an application developer, want to ask my CI/CD expert to provide me a predefined development environment, where I can have an one-liner to rebuild and immediately test the software changes I didn't even commit.


> I (really), as a CI/CD expert, product owner, and university lecturer, can only understand the requests of the second role, the app-developer role!

Maybe, if you don't understand, don't assert that this issue is useless? 

You're free to ignore it and continue using `kind load image-archive` if that's sufficient for you. 
It is *not* equivalent though, and there are valid reasons to want a UX similar to `kind load docker-image` but for multiple local image sources.

Let's try to stay on topic please.
